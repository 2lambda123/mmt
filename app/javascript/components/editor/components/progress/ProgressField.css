.field-circle {
    font-size: 1.30em;
    padding: 0 0.25em 
}

.invalid-circle {
    color: rgb(231,  76,  60)
}
.not-started-required-circle {
    color: rgb(24, 156, 84)
}

.not-started-not-required-circle {
    color: rgb(140, 140, 140)
}

.pass-required-circle {
    color: rgb(24, 156, 84)
}

.pass-not-required-circle {
    color: rgb(149, 165, 166)
}

.error-circle {
    color: rgb(231,  76,  60)
}
.progressField{
    cursor: pointer;
}