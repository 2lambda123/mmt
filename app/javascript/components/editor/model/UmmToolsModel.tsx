/* eslint-disable @typescript-eslint/no-explicit-any */
import toolInformationUiSchema from '../data/configuration/uiSchemas/tools/tool_information'
import relatedUrlsUiSchema from '../data/configuration/uiSchemas/tools/related_urls'
import toolsConfiguration from '../data/configuration/uiForms/toolsConfiguration'
import organizationUiSchema from '../data/configuration/uiSchemas/tools/organization'
import UmmModel from './UmmModel'
import descriptiveKeywordsUiSchema from '../data/configuration/uiSchemas/tools/descriptive_keywords'
import potentialActionUiSchema from '../data/configuration/uiSchemas/tools/potential_action'

export default class UmmToolsModel extends UmmModel {
  get documentTypeForDisplay() {
    return 'Tool Record'
  }
  get documentType() {
    return 'tool_drafts'
  }
  get formSections() {
    return toolsConfiguration
  }

  get uiSchema() {
    const base = { 'ui:submitButtonOptions': { norender: true, submitText: 'Save' } }

    if (this.currentSection.displayName === 'Tool Information') {
      const uiSchema: any = toolInformationUiSchema
      const urlUiSchema: any = uiSchema.URL
      urlUiSchema['ui:service'] = this.service
      urlUiSchema['ui:keyword_scheme'] = 'rucontenttype'
      return { ...uiSchema, ...base }
    }
    if (this.currentSection.displayName === 'Descriptive Keywords') {
      const uiSchema = descriptiveKeywordsUiSchema
      const toolKeywords: any = uiSchema.ToolKeywords
      toolKeywords['ui:service'] = this.service
      toolKeywords['ui:keyword_scheme'] = 'science_keywords'
      return uiSchema
    }
    if (this.currentSection.displayName === 'Tool Organizations') {
      const uiSchema: any = organizationUiSchema
      const organizations: any = uiSchema.Organizations.items
      organizations['ui:service'] = this.service
      organizations['ui:keyword_scheme'] = 'providers'
      organizations['ui:keyword_scheme_column_names'] = ['short_name', 'long_name', 'url']
      return { ...uiSchema, ...base }
    }
    if (this.currentSection.displayName === 'Potential Action') {
      const uiSchema: any = potentialActionUiSchema
      return { ...uiSchema, ...base }
    }
    if (this.currentSection.displayName === 'Related URLs') {
      const uiSchema: any = relatedUrlsUiSchema
      const urlUiSchema: any = uiSchema.RelatedURLs.items
      urlUiSchema['ui:service'] = this.service
      return { ...uiSchema, ...base }
    }
    return { ...base }
  }

  // In this case the old preview sends the same section name that the react component uses (just lowercase), so
  // nothing needs to be migrated. once the old MMT preview is replaced, this 'migrate" functionality can be removed.
  migratedSectionName(sectionName: string): string {
    return sectionName
  }
}
