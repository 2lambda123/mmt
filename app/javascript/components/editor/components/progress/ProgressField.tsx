/* eslint-disable jsx-a11y/anchor-has-content */
/* eslint-disable jsx-a11y/anchor-is-valid */
/* eslint-disable jsx-a11y/control-has-associated-label */
import React, { ReactNode } from 'react'
import _ from 'lodash'
import { FieldInfo } from './FieldInfo'
import { ProgressCircleType } from './ProgressCircleType'
import './ProgressField.css'
import withRouter from '../withRouter'
import MetadataEditor from '../../MetadataEditor'

type ProgressFieldProps = {
  fieldInfo: FieldInfo
  router: RouterType
  editor: MetadataEditor
}

class ProgressField extends React.Component<ProgressFieldProps, never> {
  progressFieldCircle(type: ProgressCircleType, required: boolean) {
    const { fieldInfo } = this.props
    const { name, error } = fieldInfo

    let message = name
    if (error) {
      message = `${name} - ${error.stack}`
    }
    let messageRequired = `${name} is required`
    if (error) {
      messageRequired = `${name} - ${error.stack}`
    }
    let icon: ReactNode
    switch (type) {
      case ProgressCircleType.Invalid:
        icon = <i title={message} className="eui-icon eui-fa-minus-circle invalid-circle" />
        break
      case ProgressCircleType.NotStarted:
        if (required) {
          icon = <i title={messageRequired} className="eui-icon eui-required-o not-started-required-circle" />
        } else {
          icon = <i title={message} className="eui-icon eui-fa-circle-o not-started-not-required-circle" />
        }
        break
      case ProgressCircleType.Pass:
        if (required) {
          icon = <i title={messageRequired} className="eui-icon eui-required icon-green pass-required-circle" />
        } else {
          icon = <i title={message} className="eui-icon eui-fa-circle icon-grey pass-not-required-circle" />
        }
        break
      case ProgressCircleType.Error:
        icon = <i title={message} className="eui-icon eui-fa-minus-circle error-circle" />
        break
      default:
        icon = <span>Status Unknown</span>
    }
    return (
      // Todo: Fix this
      // eslint-disable-next-line jsx-a11y/click-events-have-key-events, jsx-a11y/no-static-element-interactions
      <span
        className="progressField"
        onClick={() => this.navigateTo()}
        data-testid={`progress-field__${_.kebabCase(name)}_link`}
      >
        {icon}
      </span>
    )
  }
  navigateTo() {
    const { fieldInfo, router, editor } = this.props
    const {
      index, section
    } = fieldInfo
    const { params } = router
    const { id } = params
    const sectionName = section.replace(/\s/g, '_')
    if (index) {
      // editor.setArrayField(index)
    }
    setTimeout(() => {
      window.location.href = `/${editor.documentType}/${id}/edit/${sectionName}`
    }, 50)
    // navigate(`/${editor.documentType}/${id}/edit/${sectionName}`, { replace: false })
  }
  render() {
    const { fieldInfo } = this.props
    const {
      name, index = '', required, status
    } = fieldInfo
    const requiredAsString = required ? 'required' : 'not-required'
    return <span className="field-circle" key={JSON.stringify(fieldInfo)} data-testid={`progress-field__${_.kebabCase(name)}_${index}_${requiredAsString}_${_.kebabCase(status)}`}>{this.progressFieldCircle(fieldInfo.status, fieldInfo.required)}</span>
  }
}
export default (withRouter(ProgressField))
