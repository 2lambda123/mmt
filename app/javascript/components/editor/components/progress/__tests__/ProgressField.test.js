import React from 'react'
import {
  render, fireEvent
} from '@testing-library/react'
import {
  BrowserRouter
} from 'react-router-dom'
import ProgressField from '../ProgressField'
import MetadataEditor from '../../../MetadataEditor'
import UmmToolsModel from '../../../model/UmmToolsModel'

describe('Progress Field Test', () => {
  test('renders in case field required and passed', async () => {
    const props = {
      fieldInfo: {
        section: 'Tool Information',
        name: 'Name',
        index: null,
        status: 'Pass',
        required: true
      }
    }
    const { getByTestId, container } = render(
      <BrowserRouter>
        <ProgressField {...props} />
      </BrowserRouter>
    )
    const component = getByTestId('progress-field__name_null_required_pass')
    expect(component).toBeDefined()
    expect(component).not.toBeNull()
    expect(container).toMatchSnapshot()
  })
  test('renders in case field required and error', async () => {
    const props = {
      fieldInfo: {
        section: 'Tool Information',
        name: 'Name',
        index: null,
        status: 'Error',
        required: true
      }
    }
    const { getByTestId, container } = render(
      <BrowserRouter>
        <ProgressField {...props} />
      </BrowserRouter>
    )
    const component = getByTestId('progress-field__name_null_required_error')
    expect(component).toBeDefined()
    expect(component).not.toBeNull()
    expect(container).toMatchSnapshot()
  })
  test('renders in case field required and empty', async () => {
    const props = {
      fieldInfo: {
        section: 'Tool Information',
        name: 'Name',
        index: null,
        status: 'Not Started',
        required: true
      }
    }
    const { getByTestId, container } = render(
      <BrowserRouter>
        <ProgressField {...props} />
      </BrowserRouter>
    )
    const component = getByTestId('progress-field__name_null_required_not-started')
    expect(component).toBeDefined()
    expect(component).not.toBeNull()
    expect(container).toMatchSnapshot()
  })
  test('renders in case field required and invalid', async () => {
    const props = {
      fieldInfo: {
        section: 'Tool Information',
        name: 'Name',
        index: null,
        status: 'Invalid',
        required: true
      }
    }
    const { getByTestId, container } = render(
      <BrowserRouter>
        <ProgressField {...props} />
      </BrowserRouter>
    )
    const component = getByTestId('progress-field__name_null_required_invalid')
    expect(component).toBeDefined()
    expect(component).not.toBeNull()
    expect(container).toMatchSnapshot()
  })
  test('renders in case field not required and passed', async () => {
    const props = {
      fieldInfo: {
        section: 'Tool Information',
        name: 'Name',
        index: null,
        status: 'Pass',
        required: false
      }
    }
    const { getByTestId, container } = render(
      <BrowserRouter>
        <ProgressField {...props} />
      </BrowserRouter>
    )
    const component = getByTestId('progress-field__name_null_not-required_pass')
    expect(component).toBeDefined()
    expect(component).not.toBeNull()
    expect(container).toMatchSnapshot()
  })
  test('renders in case field not required and error', async () => {
    const props = {
      fieldInfo: {
        section: 'Tool Information',
        name: 'Name',
        index: null,
        status: 'Error',
        required: false
      }
    }
    const { getByTestId, container } = render(
      <BrowserRouter>
        <ProgressField {...props} />
      </BrowserRouter>
    )
    const component = getByTestId('progress-field__name_null_not-required_error')
    expect(component).toBeDefined()
    expect(component).not.toBeNull()
    expect(container).toMatchSnapshot()
  })
  test('renders in case field not required and invalid', async () => {
    const props = {
      fieldInfo: {
        section: 'Tool Information',
        name: 'Name',
        index: null,
        status: 'Invalid',
        required: false
      }
    }
    const { getByTestId, container } = render(
      <BrowserRouter>
        <ProgressField {...props} />
      </BrowserRouter>
    )
    const component = getByTestId('progress-field__name_null_not-required_invalid')
    expect(component).toBeDefined()
    expect(component).not.toBeNull()
    expect(container).toMatchSnapshot()
  })
  test('renders in case field not required and empty', async () => {
    const props = {
      fieldInfo: {
        section: 'Tool Information',
        name: 'Name',
        index: null,
        status: 'Not Started',
        required: false
      }
    }
    const { getByTestId, container } = render(
      <BrowserRouter>
        <ProgressField {...props} />
      </BrowserRouter>
    )
    const component = getByTestId('progress-field__name_null_not-required_not-started')
    expect(component).toBeDefined()
    expect(component).not.toBeNull()
    expect(container).toMatchSnapshot()
  })
  test('renders in case field status unknown', async () => {
    const props = {
      fieldInfo: {
        section: 'Tool Information',
        name: 'Name',
        index: null,
        status: 'Made Up',
        required: false
      }
    }
    const { getByTestId, container } = render(
      <BrowserRouter>
        <ProgressField {...props} />
      </BrowserRouter>
    )
    const component = getByTestId('progress-field__name_null_not-required_made-up')
    expect(component).toBeDefined()
    expect(component).not.toBeNull()
    expect(container).toMatchSnapshot()
  })
})
describe('Progress field navigation test single field', () => {
  test('navigate to edit a single field', async () => {
    const model = new UmmToolsModel()
    const editor = new MetadataEditor(model)
    const props = {
      fieldInfo: {
        section: 'Tool Information',
        name: 'Name',
        index: null,
        status: 'Pass',
        required: true
      },
      editor

    }
    const { getByTestId, queryByTestId, container } = render(
      <BrowserRouter>
        <ProgressField {...props} />
      </BrowserRouter>
    )
    const spy = jest.spyOn(ProgressField.WrappedComponent.prototype, 'navigateTo')
    const component = getByTestId('progress-field__name_null_required_pass')
    const clickField = queryByTestId('progress-field__name_link')
    fireEvent.click(clickField)
    expect(spy).toHaveBeenCalledTimes(1)
    expect(component).toBeDefined()
    expect(component).not.toBeNull()
    expect(container).toMatchSnapshot()
  })
})
