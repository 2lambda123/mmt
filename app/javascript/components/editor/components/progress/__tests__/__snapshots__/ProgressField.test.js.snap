// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Progress Field Test renders in case field not required and empty 1`] = `
<div>
  <span
    class="field-circle"
    data-testid="progress-field__name_null_not-required_not-started"
  >
    <span
      class="progressField"
      data-testid="progress-field__name_link"
    >
      <i
        class="eui-icon eui-fa-circle-o not-started-not-required-circle"
        title="Name"
      />
    </span>
  </span>
</div>
`;

exports[`Progress Field Test renders in case field not required and error 1`] = `
<div>
  <span
    class="field-circle"
    data-testid="progress-field__name_null_not-required_error"
  >
    <span
      class="progressField"
      data-testid="progress-field__name_link"
    >
      <i
        class="eui-icon eui-fa-minus-circle error-circle"
        title="Name"
      />
    </span>
  </span>
</div>
`;

exports[`Progress Field Test renders in case field not required and invalid 1`] = `
<div>
  <span
    class="field-circle"
    data-testid="progress-field__name_null_not-required_invalid"
  >
    <span
      class="progressField"
      data-testid="progress-field__name_link"
    >
      <i
        class="eui-icon eui-fa-minus-circle invalid-circle"
        title="Name"
      />
    </span>
  </span>
</div>
`;

exports[`Progress Field Test renders in case field not required and passed 1`] = `
<div>
  <span
    class="field-circle"
    data-testid="progress-field__name_null_not-required_pass"
  >
    <span
      class="progressField"
      data-testid="progress-field__name_link"
    >
      <i
        class="eui-icon eui-fa-circle icon-grey pass-not-required-circle"
        title="Name"
      />
    </span>
  </span>
</div>
`;

exports[`Progress Field Test renders in case field required and empty 1`] = `
<div>
  <span
    class="field-circle"
    data-testid="progress-field__name_null_required_not-started"
  >
    <span
      class="progressField"
      data-testid="progress-field__name_link"
    >
      <i
        class="eui-icon eui-required-o not-started-required-circle"
        title="Name"
      />
    </span>
  </span>
</div>
`;

exports[`Progress Field Test renders in case field required and error 1`] = `
<div>
  <span
    class="field-circle"
    data-testid="progress-field__name_null_required_error"
  >
    <span
      class="progressField"
      data-testid="progress-field__name_link"
    >
      <i
        class="eui-icon eui-fa-minus-circle error-circle"
        title="Name"
      />
    </span>
  </span>
</div>
`;

exports[`Progress Field Test renders in case field required and invalid 1`] = `
<div>
  <span
    class="field-circle"
    data-testid="progress-field__name_null_required_invalid"
  >
    <span
      class="progressField"
      data-testid="progress-field__name_link"
    >
      <i
        class="eui-icon eui-fa-minus-circle invalid-circle"
        title="Name"
      />
    </span>
  </span>
</div>
`;

exports[`Progress Field Test renders in case field required and passed 1`] = `
<div>
  <span
    class="field-circle"
    data-testid="progress-field__name_null_required_pass"
  >
    <span
      class="progressField"
      data-testid="progress-field__name_link"
    >
      <i
        class="eui-icon eui-required icon-green pass-required-circle"
        title="Name - Required field complete"
      />
    </span>
  </span>
</div>
`;

exports[`Progress Field Test renders in case field status unknown 1`] = `
<div>
  <span
    class="field-circle"
    data-testid="progress-field__name_null_not-required_made-up"
  >
    <span
      class="progressField"
      data-testid="progress-field__name_link"
    >
      <span>
        Status Unknown
      </span>
    </span>
  </span>
</div>
`;

exports[`Progress field navigation test single field navigate to edit a single field 1`] = `
<div>
  <span
    class="field-circle"
    data-testid="progress-field__name_null_required_pass"
  >
    <span
      class="progressField"
      data-testid="progress-field__name_link"
    >
      <i
        class="eui-icon eui-required icon-green pass-required-circle"
        title="Name - Required field complete"
      />
    </span>
  </span>
</div>
`;
